core/internal/api/credentials.go

```diff
@@ -55,3 +65,237 @@ func (c *apiKeyCredentialProvider) Apply(req *http.Request) error {
+	ExpiresAt ExpiresAt `json:"expires_at"`
+
+	// The access token to use for authentication
+	AccessToken string `json:"access_token"`
+}
+
+func (c *tokenInfo) IsTokenExpiring() bool {
+	return time.Until(time.Time(c.ExpiresAt)) <= time.Minute*5
+}
+
+// CredentialsFile is used when serializing/deserializing JSON data from the
+// credentials file
+type CredentialsFile struct {
+	Credentials map[string]tokenInfo `json:"credentials"`
+}
+
+// Apply checks if the access token is expiring, and fetches a new one if so.
+// It then supplies the access token to the request via the Authorization header
+// as a Bearer token
+func (c *oauth2CredentialProvider) Apply(req *http.Request) error {
+	if c.token.IsTokenExpiring() {
+		if err := c.loadCredentials(); err != nil {
+			return err
+		}
+	}
+	req.Header.Set(
+		"Authorization",
+		"Bearer "+c.token.AccessToken,
+	)
+	return nil
+}
+
+// loadCredentials attempts to load an access token from the credentials file.
+// If the credentials file does not exist, it creates it.
+func (c *oauth2CredentialProvider) loadCredentials() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	_, err := os.Stat(c.credentialsFilePath)
+	if os.IsNotExist(err) {
+		err = c.writeCredentialsFile()
+	}
+	if err != nil {
+		return err
+	}
+
+	return c.loadCredentialsFromFile()
+}
+
+// loadCredentialsFromFile loads the access token from the credentials file. If
+// the access token does not exist for the given base url, or the token is
+// expiring, it fetches a new one from the server, and saves it to the credentials file
+func (c *oauth2CredentialProvider) loadCredentialsFromFile() error {
+	file, err := os.ReadFile(c.credentialsFilePath)
+	if err != nil {
+		return fmt.Errorf("failed to read credentials file: %v", err)
+	}
+
+	var credsFile CredentialsFile
+	if err := json.Unmarshal(file, &credsFile); err != nil {
+		return fmt.Errorf("failed to read credentials file: %v", err)
+	}
+
+	if credsFile.Credentials == nil {
+		credsFile.Credentials = make(map[string]tokenInfo)
+	}
+
+	creds, ok := credsFile.Credentials[c.baseURL]
+
+	if !ok || creds.IsTokenExpiring() {
+		newCreds, err := c.createAccessToken()
+		if err != nil {
+			return err
+		}
+		credsFile.Credentials[c.baseURL] = *newCreds
+		updatedFile, err := json.MarshalIndent(credsFile, "", "  ")
+		if err != nil {
+			return fmt.Errorf("failed to update credentials file: %v", err)
+		}
+		err = os.WriteFile(c.credentialsFilePath, updatedFile, 0600)
+		if err != nil {
+			return fmt.Errorf("failed to update credentials file: %v", err)
+		}
+		c.token = *newCreds
+	} else {
+		c.token = creds
+	}
+
+	return nil
+}
+
+// writeCredentialsFile obtains an access token from the server and saves it
+// to a credentials file
+func (c *oauth2CredentialProvider) writeCredentialsFile() error {
+	token, err := c.createAccessToken()
+	if err != nil {
+		return err
+	}
+
+	data := CredentialsFile{
+		Credentials: map[string]tokenInfo{
+			c.baseURL: *token,
```

Comment:
You can avoid dereferencing if you always pass `token` by value, which would be a lot less error-prone!

Reply:
good call, done